File: pyqt/src/widgets/board_widget.py
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5.QtCore import *
import copy
from .square_widget import SquareWidget

# controllers
import os
import sys
sys.path.append('{}/../'.format(os.path.dirname(os.path.abspath(__file__))))
from controllers.rush_hour import RushHour


class BoardWidget(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)

        # data
        self.carSelected = None

        # widgets
        self.squares = [[SquareWidget(i, j) for j in range(6)] for i in range(6)]

        # layouts
        self.gridLayout = QGridLayout()
        self.gridLayout.setSpacing(0)

        for i in range(6):
            for j in range(6):
                self.gridLayout.addWidget(self.squares[i][j], i, j)
        
        self.mainLayout = QVBoxLayout()
        self.mainLayout.addWidget(QLabel("Use arrow keys to move cars"))
        self.mainLayout.addLayout(self.gridLayout)

        self.setLayout(self.mainLayout)

    def renderLevel(self, level, cars):
        # clear selection
        self.carSelected = None

        for i in range(6):
            for j in range(6):
                self.squares[i][j].setText(level[i][j])
                if level[i][j] in cars:
                    car = cars[level[i][j]]
                    style = f'background: {car["color"]}; border: 3px solid black;'
                    self.squares[i][j].setStyleSheet(style)
                else:
                    self.squares[i][j].setStyleSheet("border: 3px solid black;")

    def onSquareClicked(self, square):
        self.parent().onSquareClicked(square)

    def selectCar(self, car):
        if self.carSelected:
            # unselect previous car
            for square in self.carSelected["squares"]:
                self.squares[square[0]][square[1]].updateAlpha(255)
        
        # select new car
        for square in car["squares"]:
            self.squares[square[0]][square[1]].updateAlpha(50)
        
        self.carSelected = car

    def moveSelectedCar(self, board, key):
        """
        board: [['.', '.', 'A', 'A', '.', '.'], ...]
        key: 'Up', 'Down', 'Left' or 'Right'
        """
        init_board = copy.deepcopy(board)

        controller = RushHour(board)
        carName = self.carSelected["name"]
        dir = RushHour.UP if key == 'Up' \
            else RushHour.DOWN if key == 'Down' \
            else RushHour.LEFT if key == 'Left' \
            else RushHour.RIGHT if key == 'Right' \
            else None
        controller.make_move(carName, dir, length=1)

        return init_board != board

File: wasm/rushhour/src/widgets/boardWidget.cpp
#include "boardWidget.h"
#include "mainWidget.h"
#include "squareWidget.h"
#include "../controllers/rushHour.h"

BoardWidget::BoardWidget(MainWidget *mainWidget, QWidget *parent) :
    mainWidget(mainWidget),
    QWidget(parent)
{
    // data
    carSelected.clear();

    // widgets
    for (int i = 0; i < 6; i++)
    {
        squares.push_back(std::vector<SquareWidget*>(6));
        for (int j = 0; j < 6; j++)
        {
            squares[i][j] = new SquareWidget(i, j, this);
        }
    }

    // layouts
    gridLayout = new QGridLayout();
    gridLayout->setSpacing(0);

    for(int i = 0; i < 6; i++) {
        for(int j = 0; j < 6; j++) {
            gridLayout->addWidget(squares[i][j], i, j);
        }
    }

    mainLayout = new QVBoxLayout();
    //mainLayout->addWidget(instructionLabel);
    mainLayout->addLayout(gridLayout);

    setLayout(mainLayout);
}

const Car &BoardWidget::getCarSelected()
{
    return carSelected;
}

void BoardWidget::setCarSelected(const Car &pCarSelected)
{
    carSelected = pCarSelected;
}

void BoardWidget::renderLevel(const std::vector<std::vector<char>> &level,
                              const std::map<char, Car> &cars)
{
    // clear selection
    carSelected.clear();

    for (int i = 0; i < 6; i++)
    {
        for (int j = 0; j < 6; j++)
        {
            squares[i][j]->setText(QString::fromStdString(std::string(1, level[i][j])));
            if (cars.find(level[i][j]) != cars.end())
            {
                auto carName = level[i][j];
                auto car = cars.find(carName)->second;
                std::string style = "background: " + car.color + "; border: 3px solid black;";
                squares[i][j]->setStyleSheet(QString::fromStdString(style));
            }
            else
            {
                squares[i][j]->setStyleSheet("border: 3px solid black;");
            }
        }
    }
}

void BoardWidget::onSquareClicked(SquareWidget *square)
{
    mainWidget->onSquareClicked(square);
}

void BoardWidget::selectCar(const Car &car)
{
    // unselect previous car
    for (auto square : carSelected.squares)
    {
        squares[square.first][square.second]->updateAlpha(255);
    }

    // select new car
    for (auto square : car.squares)
    {
        squares[square.first][square.second]->updateAlpha(50);
    }

    carSelected = car;
}

bool BoardWidget::moveSelectedCar(std::vector<std::vector<char>> &board, std::string key)
{
    std::vector<std::vector<char>> init_board = board;

    auto controller = RushHour(board);
    char carName = carSelected.name;
    int dir = (key == "Up") ? RushHour::UP : (key == "Down") ? RushHour::DOWN : (key == "Left") ? 
        RushHour::LEFT : (key == "Right") ? RushHour::RIGHT : -1;

    controller.makeMove(carName, dir, 1);

    return init_board != board;
}